### PR TITLE
RFC(react): Allow returning string, boolean etc from FunctionComponent

### LINKED ISSUE #18051: React components should return ReactNode, or children should be ReactElement
- [x] I tried using the `@types/react` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @ericanderson @onigoetz @tkrotoff @digiguru @morcerf @johnnyreilly @bbenezech @pzavolinsky @DovydasNavickas 

The following components 

```
const SFC: React.StatelessComponent = (props) => props.children;

class C extends React.Component {
  render() {
    return this.props.children;
  }
}
```

give the errors

```
Type '(props: { children?: ReactNode; }) => ReactNode' is not assignable to type 'StatelessComponent<{}>'.
  Type 'ReactNode' is not assignable to type 'ReactElement<any> | null'.
    Type 'undefined' is not assignable to type 'ReactElement<any> | null'.

Class 'C' incorrectly extends base class 'Component<{}, {}>'.
  Types of property 'render' are incompatible.
    Type '() => ReactNode' is not assignable to type '() => false | Element | null'.
      Type 'ReactNode' is not assignable to type 'false | Element | null'.
        Type 'undefined' is not assignable to type 'false | Element | null'.
```

It works if you wrap `children` in an `Element`.

Returning `children` is quite common when making Provider components that just add something to `context`.

### LINKED ISSUE #18912: React typings allow false for class components, but not for stateless
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @gaspard @vbfox 

As mentioned in [React documentation](https://facebook.github.io/react/docs/conditional-rendering.html#inline-if-with-logical--operator), a stateless component can return `false` for empty content. Currently [StatelessComponent](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L312-L313) is defined as:
```ts
interface StatelessComponent<P = {}> {
        (props: P & { children?: ReactNode }, context?: any): ReactElement<any> | null;
```
whereas [Component](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L270-L280) has
```ts
class Component<P, S> {
        constructor(props?: P, context?: any);
        // ...snip...
        render(): JSX.Element | null | false;
```

Shouldn't these types be the same (including eliminating the ReactElement vs JSX.Element difference)? And to avoid future such issues (#14064, this one) it should probably be a named type, e.g. RenderResult.

A search for `| null` also comes up with `ComponentSpec.render` which probably should also have the same result type.

### PR DESCRIPTION
**Pull request by eps1lon**
Hack to fix all the "why can't I return X from a function component" issues. We would no longer need a fix for https://github.com/microsoft/TypeScript/issues/14729 or https://github.com/microsoft/TypeScript/issues/21699.

Only worked on the React type tests for now since it highlights the issues we would have with this change.

The problem with this is that you can no longer enforce receiving actual JSX elements (the ones with `{ props: P, type: T, key: string }`). `SuspenseList` is an example of a built-in component that relies on that.

In other words, we basically disable type-safe introspection of JSX elements.

Closes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18051
Closes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18912

### CODE REVIEW

File: types/react/index.d.ts [hunk 1]
```diff
@@ -3119,9 +3119,9 @@ type ReactManagedAttributes<C, P> = C extends { propTypes: infer T; defaultProps
 
 declare global {
     namespace JSX {
-        interface Element extends React.ReactElement<any, any> { }
+        type Element = React.ReactNode;
```
Comment (eps1lon) [inline types/react/index.d.ts:3122]: This is the core of the change. It's a hack but it trades the linked issues with new ones (credit to @sebmarkbage).

Comment (sebmarkbage) [reply types/react/index.d.ts:3122]: I think the main issue is things like:

```
function Foo() : React.ReactElement {
  return <div />;
}
```

It's not wrong to type it this way since this function will always return an actual element but it would error in a lot of places.
